File: src/web_optimise/_display.py
"""Rich console display functions for CLI output."""

from typing import TYPE_CHECKING
from typing import Final

from rich.console import Console
from rich.table import Table

from web_optimise._constants import BYTES_PER_MB
from web_optimise._constants import MKV_EXTENSIONS
from web_optimise._discovery import effective_mode
from web_optimise._discovery import sanitise_filename
from web_optimise._modes import MODE_CONFIGS

if TYPE_CHECKING:
    from web_optimise._constants import Mode
    from web_optimise._types import EncodeResult
    from web_optimise._types import FileInfo

console: Final[Console] = Console()


def print_summary(results: tuple[EncodeResult, ...], /) -> None:
    """Print before/after file size comparison table."""
    table = Table(title="Encoding Results")
    table.add_column("File", style="cyan")
    table.add_column("Output", style="dim")
    table.add_column("Input Size", justify="right")
    table.add_column("Output Size", justify="right")
    table.add_column("Change", justify="right")

    total_input = 0
    total_output = 0

    for r in results:
        total_input += r.input_size
        total_output += r.output_size
        input_mb = r.input_size / BYTES_PER_MB
        output_mb = r.output_size / BYTES_PER_MB

        style = "green" if r.savings_pct < 0 else "red"
        table.add_row(
            r.input_path.name,
            r.output_path.name,
            f"{input_mb:.1f} MB",
            f"{output_mb:.1f} MB",
            f"[{style}]{r.savings_display}[/{style}]",
        )

    _add_totals_row(table, total_input, total_output)
    console.print(table)


def _add_totals_row(
    table: Table,
    total_input: int,
    total_output: int,
) -> None:
    """Add a bold totals row to the results table."""
    if total_input == 0:
        return
    total_input_mb = total_input / BYTES_PER_MB
    total_output_mb = total_output / BYTES_PER_MB
    total_pct = ((total_output - total_input) / total_input) * 100
    sign = "+" if total_pct > 0 else ""
    style = "green" if total_pct < 0 else "red"
    table.add_row(
        "[bold]Total[/bold]",
        "",
        f"[bold]{total_input_mb:.1f} MB[/bold]",
        f"[bold]{total_output_mb:.1f} MB[/bold]",
        f"[bold][{style}]{sign}{total_pct:.1f}%[/{style}][/bold]",
    )


def display_analysis(
    infos: tuple[FileInfo, ...],
    *,
    user_mode: Mode,
) -> None:
    """Print the pre-encode file analysis table."""
    table = Table(title="File Analysis")
    table.add_column("File", style="cyan")
    table.add_column("Output Preview", style="dim")
    table.add_column("Duration", justify="right")
    table.add_column("Size", justify="right")
    table.add_column("Resolution")
    table.add_column("Codec")
    table.add_column("Video Bitrate", justify="right")

    for info in infos:
        file_mode = effective_mode(info, user_mode=user_mode)
        config = MODE_CONFIGS[file_mode]
        table.add_row(
            info.path.name,
            sanitise_filename(info.path.name, ext=config.output_ext),
            info.duration_display,
            f"{info.size_mb:.1f} MB",
            info.video.resolution_label,
            f"{info.video.codec} ({info.video.profile})",
            info.video.bitrate_kbps,
        )

    console.print(table)


def display_remux_warnings(infos: tuple[FileInfo, ...]) -> None:
    """Print advisory notes about MP4 files that may benefit from full encoding."""
    warnings: list[str] = []
    for info in infos:
        if info.path.suffix.lower() in MKV_EXTENSIONS:
            continue
        name = info.path.name
        if info.video.codec != "h264":
            warnings.append(
                f"  [yellow]![/yellow] {name}: codec is {info.video.codec} "
                f"(not H.264). Use --mode encode to re-encode."
            )
        elif "High" not in info.video.profile:
            warnings.append(
                f"  [yellow]![/yellow] {name}: profile is {info.video.profile} "
                f"(not High). Use --mode encode to upgrade."
            )
    if warnings:
        console.print()
        for w in warnings:
            console.print(w)

File: src/WebOptimise/Display.fs
namespace WebOptimise

open System.IO
open System.Threading
open System.Threading.Tasks
open Spectre.Console
open SpectreCoff

/// SpectreCoff display functions for CLI output.
[<RequireQualifiedAccess>]
module Display =

    let printError (msg: string) =
        Many [
            MC(Color.Red, "Error:")
            V $" %s{msg}"
        ]
        |> toConsole

    let displayAnalysis (infos: MediaFileInfo list) (userMode: Mode) =
        let rightAligned = { defaultColumnLayout with Alignment = Right }

        let columns = [
            column (C "File")
            column (C "Output Preview")
            column (C "Duration") |> withLayout rightAligned
            column (C "Size") |> withLayout rightAligned
            column (C "Resolution")
            column (C "Codec")
            column (C "Video Bitrate") |> withLayout rightAligned
        ]

        let rows =
            infos
            |> List.map (fun info ->
                let fileMode = Discovery.effectiveMode info userMode
                let config = ModeConfig.forMode fileMode

                Payloads [
                    V(MediaFilePath.name info.Path)
                    C(Discovery.sanitiseFilename (MediaFilePath.name info.Path) config.OutputExt)
                    V(MediaFileInfo.durationDisplay info)
                    V $"%.1f{MediaFileInfo.sizeMB info} MB"
                    V(VideoStream.resolutionLabel info.Video)
                    V(
                        VideoCodec.displayName info.Video.Codec
                        + " ("
                        + (
                            match info.Video.Profile with
                            | ValueSome p -> VideoProfile.displayName p
                            | ValueNone -> "N/A"
                        )
                        + ")"
                    )
                    V(VideoStream.bitrateKbps info.Video)
                ]
            )

        table columns rows
        |> withTitle "File Analysis"
        |> toOutputPayload
        |> toConsole

    let displayRemuxWarnings (infos: MediaFileInfo list) =
        let warnings =
            infos
            |> List.collect (fun info ->
                if Constants.mkvExtensions.Contains(MediaFilePath.extension info.Path) then
                    []
                else
                    let name = MediaFilePath.name info.Path

                    [
                        if info.Video.Codec <> VideoCodec.H264 then
                            Many [
                                MC(Color.Yellow, "!")
                                V
                                    $" %s{name}: codec is %s{VideoCodec.displayName info.Video.Codec} (not H.264). Use --mode encode to re-encode."
                            ]
                        elif info.Video.Profile <> ValueSome VideoProfile.High then
                            let profileLabel =
                                match info.Video.Profile with
                                | ValueSome p -> VideoProfile.displayName p
                                | ValueNone -> "N/A"

                            Many [
                                MC(Color.Yellow, "!")
                                V $" %s{name}: profile is %s{profileLabel} (not High). Use --mode encode to upgrade."
                            ]
                    ]
            )

        if not warnings.IsEmpty then
            NL |> toConsole
            warnings |> List.iter toConsole

    let printSummary (results: EncodeResult list) =
        let rightAligned = { defaultColumnLayout with Alignment = Right }

        let noWrap = { defaultColumnLayout with Wrap = false }

        let columns = [
            column (C "File") |> withLayout noWrap
            column (C "Output") |> withLayout noWrap
            column (C "Input Size") |> withLayout rightAligned
            column (C "Output Size") |> withLayout rightAligned
            column (C "Change") |> withLayout rightAligned
        ]

        let dataRows =
            results
            |> List.map (fun r ->
                let inputMb = float r.InputSize / float Constants.BytesPerMB
                let outputMb = float r.OutputSize / float Constants.BytesPerMB

                let color =
                    if EncodeResult.savingsPct r < 0.0 then
                        Color.Green
                    else
                        Color.Red

                Payloads [
                    V(MediaFilePath.name r.InputPath)
                    V(Path.GetFileName r.OutputPath |> NullSafe.path)
                    V $"%.1f{inputMb} MB"
                    V $"%.1f{outputMb} MB"
                    MC(color, EncodeResult.savingsDisplay r)
                ]
            )

        let struct (totalInput, totalOutput) =
            (struct (0L, 0L), results)
            ||> List.fold (fun (struct (ti, to')) r -> struct (ti + r.InputSize, to' + r.OutputSize))

        let allRows =
            if totalInput > 0L then
                let totalInputMb = float totalInput / float Constants.BytesPerMB

                let totalOutputMb =
                    float totalOutput / float Constants.BytesPerMB

                let totalPct =
                    float (totalOutput - totalInput) / float totalInput * 100.0

                let sign = if totalPct > 0.0 then "+" else ""

                let color =
                    if totalPct < 0.0 then
                        Color.Green
                    else
                        Color.Red

                let totalsRow =
                    Payloads [
                        P "Total"
                        V ""
                        P $"%.1f{totalInputMb} MB"
                        P $"%.1f{totalOutputMb} MB"
                        MC(color, $"%s{sign}%.1f{totalPct}%%")
                    ]

                dataRows @ [ totalsRow ]
            else
                dataRows

        table columns allRows
        |> withTitle "Encoding Results"
        |> toOutputPayload
        |> toConsole

    let private verifyOne (pr: ProcessResult) : Task<bool> =
        task {
            let config = ModeConfig.forMode pr.Mode
            let! verifyResult = config.Verifier pr.Result.OutputPath

            let name =
                Path.GetFileName pr.Result.OutputPath |> NullSafe.path

            match verifyResult with
            | Ok() ->
                Many [
                    MC(Color.Green, "\u2713")
                    V $" %s{name}"
                ]
                |> toConsole

                return true
            | Error issues ->
                Many [
                    MC(Color.Red, "\u2717")
                    V $" %s{name}"
                ]
                |> toConsole

                for issue in issues do
                    Many [
                        MC(Color.Red, "  \u2192")
                        V $" %s{issue}"
                    ]
                    |> toConsole

                return false
        }

    let displayVerification (results: ProcessResult list) : Task<bool> =
        task {
            P "Verifying outputs..." |> toConsole

            let! outcomes = results |> List.map verifyOne |> Task.WhenAll

            return outcomes |> Array.forall id
        }

    /// Corrected SpectreCoff.Progress.startCustom — upstream ignores AutoClear/HideCompleted.
    let private startProgress (template: ProgressTemplate) (operation: ProgressOperation<'T>) =
        AnsiConsole
            .Progress()
            .AutoClear(template.AutoClear)
            .AutoRefresh(template.AutoRefresh)
            .HideCompleted(template.HideCompleted)
            .Columns(template.Columns |> List.toArray)
            .StartAsync(operation)

    let private processWithContext
        (infos: MediaFileInfo list)
        (processOne:
            MediaFileInfo -> Mode -> (float -> unit) -> CancellationToken -> Task<Result<EncodeResult, AppError>>)
        (userMode: Mode)
        (ct: CancellationToken)
        (ctx: ProgressContext)
        : Task<Result<ProcessResult list, AppError>>
        =
        let total = infos.Length

        let rec loop (remaining: MediaFileInfo list) (acc: ProcessResult list) (idx: int) =
            task {
                match remaining with
                | [] -> return Ok(List.rev acc)
                | info :: rest ->
                    let fileMode = Discovery.effectiveMode info userMode

                    let description =
                        $"[%d{idx}/%d{total}] %s{MediaFilePath.name info.Path}"

                    let progressTask =
                        realizeIn ctx (HotCustomTask(info.DurationSecs, description))

                    let onProgress elapsed = progressTask.Value <- elapsed

                    match! processOne info fileMode onProgress ct with
                    | Ok encodeResult ->
                        progressTask.Value <- info.DurationSecs

                        return!
                            loop
                                rest
                                ({
                                    Result = encodeResult
                                    Mode = fileMode
                                 }
                                 :: acc)
                                (idx + 1)
                    | Error e -> return Error e
            }

        loop infos [] 1

    let withProgress
        (infos: MediaFileInfo list)
        (processOne:
            MediaFileInfo -> Mode -> (float -> unit) -> CancellationToken -> Task<Result<EncodeResult, AppError>>)
        (userMode: Mode)
        (ct: CancellationToken)
        : Task<Result<ProcessResult list, AppError>>
        =

        let template = {
            AutoClear = false
            AutoRefresh = true
            HideCompleted = false
            Columns = [
                SpinnerColumn()
                TaskDescriptionColumn()
                ProgressBarColumn()
                PercentageColumn()
                ElapsedTimeColumn()
                RemainingTimeColumn()
            ]
        }

        startProgress template (processWithContext infos processOne userMode ct)
